ml-agents-envs/mlagents/envs/base_env.py

```diff
@@ -0,0 +1,291 @@
+     - done is a bool. Is true if the Agent was terminated during the last
+     simulation step.
+     - max_step is a bool. Is true if the Agent reached its maximum number of
+     steps during the last simulation step.
+     - agent_id is an int and an unique identifier for the corresponding Agent.
+     - action_mask is an optional list of one dimensional array of booleans.
+     Only available in multi-discrete action space type.
+     Each array corresponds to an action branch. Each array contains a mask
+     for each action of the branch. If true, the action is not available for
+     the agent during this simulation step.
+    """
+
+    obs: List[np.array]
+    reward: float
+    done: bool
+    max_step: bool
+    agent_id: int
+    action_mask: Optional[List[np.array]]
+
+
+class BatchedStepResult(NamedTuple):
+    """
+    Contains the data a group of similar Agents collected since the last
+    simulation step. Note that all Agents do not necessarily have new
+    information to send at each simulation step. Therefore, the ordering of
+    agents and the batch size of the BatchedStepResult are not fixed across
+    simulation steps.
+     - obs is a list of numpy arrays observations collected by the group of
+     agent. Each obs has one extra dimension compared to StepResult: the first
+     dimension of the array corresponds to the batch size of
+     the group.
+     - reward is a float vector of length batch size. Corresponds to the
+     rewards collected by each agent since the last simulation step.
+     - done is an array of booleans of length batch size. Is true if the
+     associated Agent was terminated during the last simulation step.
+     - max_step is an array of booleans of length batch size. Is true if the
+     associated Agent reached its maximum number of steps during the last
+     simulation step.
+     - agent_id is an int vector of length batch size containing unique
+     identifier for the corresponding Agent. This is used to track Agents
+     across simulation steps.
+     - action_mask is an optional list of two dimensional array of booleans.
+     Only available in multi-discrete action space type.
+     Each array corresponds to an action branch. The first dimension of each
+     array is the batch size and the second contains a mask for each action of
+     the branch. If true, the action is not available for the agent during
+     this simulation step.
+    """
+
+    obs: List[np.array]
+    reward: np.array
+    done: np.array
+    max_step: np.array
+    agent_id: np.array
+    action_mask: Optional[List[np.array]]
+
+    def get_agent_step_result(self, agent_id: int) -> StepResult:
+        """
+        returns the step result for a specific agent.
+        :param agent_id: The id of the agent
+        :returns: obs, reward, done, agent_id and optional action mask for a
+        specific agent
+        """
+        try:
+            agent_index = np.where(self.agent_id == agent_id)[0][0]
```

Comment:
What do you think about having an agent_id -> index dictionary, but only building it the first time get_agent_step_result() is called? I'm worried that this is O(num agents) per call, and if you're using this once, you'll probably use it (num agents) times.

Reply:
I am unsure since `BatchedStepResponse` is a namedtuple, I do not think I can have references to dicts in there.

Reply:
You definitely can have a reference (it's python! everything is a reference!) but I forget how to have a default value like this. If you think it would be useful I'll look into it.

Reply:
Made it an object for now. We can see if we really need it to be a named tuple.

Reply:
Thanks. If you really want to prematurely optimize things, you can start adding `__slots__` to classes...

Reply:
Also https://stackoverflow.com/questions/59205528/typing-namedtuple-with-a-dictionary-default